File: backend/report.js
const fs = require('fs');

const MAX_TOP_SONGS = 10;




async function agregate_artists(tracks) {
    let artists_count = {};
    let artists_playcount = {};
    let artists_lenght = {};

    tracks.forEach(track => {
        track.artists.forEach(artist => {
            const artistName = artist.name;
            artists_count[artistName] = (artists_count[artistName] || 0) + 1;
            artists_playcount[artistName] = (artists_playcount[artistName] || 0) + track.playcount;
            artists_lenght[artistName] = (artists_lenght[artistName] || 0) + track.duration_ms;
        });
    });

    let artists_count_array = Object.entries(artists_count).map(([name, count]) => ({ name, count }));
    artists_count_array.sort((a, b) => b.count - a.count)
        .splice(MAX_TOP_SONGS);

    let artists_playcount_array = Object.entries(artists_playcount).map(([name, count]) => ({ name, count }));
    artists_playcount_array.sort((a, b) => b.count - a.count)
        .splice(MAX_TOP_SONGS);

    return [artists_count_array, artists_playcount_array];
}


async function agregate_dates(tracks) {
    const sortedTracks = tracks
        .map(track => {
            const { name, artists, playcount, release_date, release_date_precision } = track;

            let normalizedDate = release_date;
            if (release_date_precision === "month") {
                normalizedDate = `${release_date}-01`;
            } 
            else if (release_date_precision === "year") {
                normalizedDate = `${release_date}-01-01`;
            }

            return {
                name,
                artists: artists.map(artist => artist.name),
                playcount,
                release_date: normalizedDate,
            };
        })
        .sort((a, b) => a.release_date.localeCompare(b.release_date));

    const oldest_tracks = sortedTracks.slice(0, MAX_TOP_SONGS);
    const newest_tracks = sortedTracks.slice(-MAX_TOP_SONGS).reverse();

    return [oldest_tracks, newest_tracks];
}


async function agregate_playcount(tracks) {
    const sorted_tracks = tracks.map(track => {
                                    const { name, artists, playcount } = track;
                                    return {
                                        name,
                                        artists: artists.map(artist => artist.name),
                                        playcount,
                                    };
                                })
                                .sort((a, b) => b.playcount - a.playcount);
    
    const most_streamed_tracks = sorted_tracks.slice(0, MAX_TOP_SONGS);
    const least_streamed_tracks = sorted_tracks.slice(-MAX_TOP_SONGS).reverse();

    return [most_streamed_tracks, least_streamed_tracks];
}


async function agregate_popularity(tracks) {
    const sorted_tracks = tracks.map(track => {
                                    const { name, artists, popularity, playcount } = track;
                                    return {
                                        name,
                                        artists: artists.map(artist => artist.name),
                                        popularity,
                                        playcount,
                                    };
                                })
                                .sort((a, b) => {
                                    if (b.popularity !== a.popularity) {
                                        return b.popularity - a.popularity;
                                    }
                                    return b.playcount - a.playcount;
                                });
    
    const most_popular_tracks = sorted_tracks.slice(0, MAX_TOP_SONGS);
    const least_popular_tracks = sorted_tracks.slice(-MAX_TOP_SONGS).reverse();

    return [most_popular_tracks, least_popular_tracks];
}


async function agregate_lenght(tracks) {
    const sorted_tracks = tracks.map(track => {
                                    const { name, artists, duration } = track;
                                    return {
                                        name,
                                        artists: artists.map(artist => artist.name),
                                        duration,
                                    };
                                })
                                .sort((a, b) => b.duration - a.duration );
    
    const most_long_tracks = sorted_tracks.slice(0, MAX_TOP_SONGS);
    const most_short_tracks = sorted_tracks.slice(-MAX_TOP_SONGS).reverse();

    return [most_long_tracks, most_short_tracks];
}



async function generate_report(playlistFile, outputReportFile) {
    const playlist = JSON.parse(fs.readFileSync(playlistFile, 'utf-8'));
    const tracks = playlist.items;

    const [artists_counts, artists_playcounts] = await agregate_artists(tracks);
    const [oldest_tracks, newest_tracks] = await agregate_dates(tracks);
    const [most_streamed_tracks, least_streamed_tracks] = await agregate_playcount(tracks);
    const [most_popular_tracks, least_popular_tracks] = await agregate_popularity(tracks);
    const [most_long_tracks, most_short_tracks] = await agregate_lenght(tracks);

    final_report = {
        total_tracks: tracks.length,
        total_artists: artists_counts.length,
        total_streams: tracks.reduce((acc, track) => acc + track.playcount, 0),
        artists_counts: artists_counts,
        artists_playcounts: artists_playcounts,
        oldest_tracks: oldest_tracks,
        newest_tracks: newest_tracks,
        most_streamed_tracks: most_streamed_tracks,
        least_streamed_tracks: least_streamed_tracks,
        most_popular_tracks: most_popular_tracks,
        least_popular_tracks: least_popular_tracks,
        most_long_tracks: most_long_tracks,
        most_short_tracks: most_short_tracks,
    }

    fs.writeFileSync(outputReportFile, JSON.stringify(final_report, null, 4), 'utf-8');
}




module.exports = { generate_report };

File: report.py
from collections import defaultdict
import json

MAX_TOP_SONGS = 10




def aggregate_general(tracks):
    total_tracks = len(tracks)
    total_artists = len(set(artist['name'] for track in tracks for artist in track['artists']))
    total_streams = sum(track['playcount'] for track in tracks)
    return total_tracks, total_artists, total_streams


def aggregate_artists(tracks):
    artists_count = defaultdict(int)
    artists_playcount = defaultdict(int)
    artists_length = defaultdict(int)

    for track in tracks:
        for artist in track['artists']:
            artist_name = artist['name']
            artists_count[artist_name] += 1
            artists_playcount[artist_name] += track['playcount']
            artists_length[artist_name] += track['duration_ms']

    artists_count_array = sorted(artists_count.items(), key=lambda x: x[1], reverse=True)[:MAX_TOP_SONGS]
    artists_playcount_array = sorted(artists_playcount.items(), key=lambda x: x[1], reverse=True)[:MAX_TOP_SONGS]
    artists_length_array = sorted(artists_length.items(), key=lambda x: x[1], reverse=True)[:MAX_TOP_SONGS]

    return artists_count_array, artists_playcount_array, artists_length_array


def aggregate_dates(tracks):
    normalized_tracks = []
    for track in tracks:
        release_date = track['release_date']
        precision = track['release_date_precision']

        if precision == "month":
            release_date += "-01"
        elif precision == "year":
            release_date += "-01-01"

        normalized_tracks.append({
            "name": track['name'],
            "artists": [artist['name'] for artist in track['artists']],
            "playcount": track['playcount'],
            "release_date": release_date
        })

    sorted_tracks = sorted(normalized_tracks, key=lambda x: x['release_date'])
    oldest_tracks = sorted_tracks[:MAX_TOP_SONGS]
    newest_tracks = sorted_tracks[-MAX_TOP_SONGS:][::-1]

    return oldest_tracks, newest_tracks


def aggregate_playcount(tracks):
    sorted_tracks = sorted(tracks, key=lambda x: x['playcount'], reverse=True)

    most_streamed_tracks = sorted_tracks[:MAX_TOP_SONGS]
    least_streamed_tracks = sorted_tracks[-MAX_TOP_SONGS:][::-1]

    return most_streamed_tracks, least_streamed_tracks


def aggregate_popularity(tracks):
    sorted_tracks = sorted(tracks, key=lambda x: (x['popularity'], x['playcount']), reverse=True)

    most_popular_tracks = sorted_tracks[:MAX_TOP_SONGS]
    least_popular_tracks = sorted_tracks[-MAX_TOP_SONGS:][::-1]

    return most_popular_tracks, least_popular_tracks


def aggregate_length(tracks):
    sorted_tracks = sorted(tracks, key=lambda x: x['duration_ms'], reverse=True)

    most_long_tracks = sorted_tracks[:MAX_TOP_SONGS]
    most_short_tracks = sorted_tracks[-MAX_TOP_SONGS:][::-1]

    return most_long_tracks, most_short_tracks


def aggregate_periods(tracks):
    year_count = defaultdict(int)
    month_count = defaultdict(int)

    for track in tracks:
        release_date = track['release_date']
        precision = track['release_date_precision']

        if precision == "year":
            year_count[release_date] += 1
        else:
            year, month, *_ = release_date.split("-")
            year_count[year] += 1
            if month:
                month_count[month] += 1

    return dict(year_count), dict(month_count)


def generate_report(playlist_file, output_report_file):
    with open(playlist_file, 'r', encoding='utf-8') as f:
        playlist = json.load(f)

    tracks = playlist['items']

    total_tracks, total_artists, total_streams = aggregate_general(tracks)
    artists_counts, artists_playcounts, artists_length = aggregate_artists(tracks)
    oldest_tracks, newest_tracks = aggregate_dates(tracks)
    most_streamed_tracks, least_streamed_tracks = aggregate_playcount(tracks)
    most_popular_tracks, least_popular_tracks = aggregate_popularity(tracks)
    most_long_tracks, most_short_tracks = aggregate_length(tracks)
    year_count, month_count = aggregate_periods(tracks)

    final_report = {
        "name": playlist['name'],
        "description": playlist['description'],
        "generatedTimeStamp": playlist['generatedTimeStamp'],
        "followers": playlist['followers'],
        "coverUrl": playlist['coverUrl'],
        "coverHex": playlist['coverHex'],
        "total_tracks": total_tracks,
        "total_artists": total_artists,
        "total_streams": total_streams,
        "artists_counts": artists_counts,
        "artists_playcounts": artists_playcounts,
        "artists_length": artists_length,
        "oldest_tracks": oldest_tracks,
        "newest_tracks": newest_tracks,
        "most_streamed_tracks": most_streamed_tracks,
        "least_streamed_tracks": least_streamed_tracks,
        "most_popular_tracks": most_popular_tracks,
        "least_popular_tracks": least_popular_tracks,
        "most_long_tracks": most_long_tracks,
        "most_short_tracks": most_short_tracks,
        "year_count": year_count,
        "month_count": month_count
    }

    with open(output_report_file, 'w', encoding='utf-8') as f:
        json.dump(final_report, f, ensure_ascii=False, indent=4)